core/trino-main/src/main/java/io/trino/execution/SetTimeZoneTask.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Override
+    public String getName()
+    {
+        return "SET TIME ZONE";
+    }
+
+    @Override
+    public ListenableFuture<?> execute(
+            SetTimeZone statement,
+            TransactionManager transactionManager,
+            Metadata metadata,
+            AccessControl accessControl,
+            QueryStateMachine stateMachine,
+            List<Expression> parameters,
+            WarningCollector warningCollector)
+    {
+        TimeZoneKey timeZoneKey;
+        if (statement.getTimeZone().isPresent()) {
+            Object timeZoneValue = evaluate(
+                    statement.getTimeZone().get(),
+                    statement,
+                    metadata,
+                    accessControl,
+                    stateMachine,
+                    parameters,
+                    warningCollector);
+
+            if (timeZoneValue instanceof Slice) {
+                timeZoneKey = getTimeZoneKey(((Slice) timeZoneValue).toStringUtf8());
+            }
+            else if (timeZoneValue instanceof Long) {
+                timeZoneKey = getTimeZoneKeyForOffset(getZoneOffsetMinutes((Long) timeZoneValue));
+            }
+            else {
+                throw new TrinoException(INVALID_LITERAL, "Time Zone value not supported");
+            }
+        }
+        else {
+            timeZoneKey = TimeZoneKey.getTimeZoneKey(TimeZone.getDefault().getID());
+        }
+        String zoneId = timeZoneKey.getId();
+        stateMachine.addSetSessionProperties(TIME_ZONE_ID, zoneId);
+
+        return immediateFuture(null);
+    }
+
+    private Object evaluate(
+            Expression expression,
+            SetTimeZone statement,
+            Metadata metadata,
+            AccessControl accessControl,
+            QueryStateMachine stateMachine,
+            List<Expression> parameters,
+            WarningCollector warningCollector)
+    {
+        Session session = stateMachine.getSession();
+        Analysis analysis = new Analyzer(session, metadata, sqlParser, groupProvider, accessControl, Optional.empty(), parameters,
+                parameterExtractor(statement, parameters), stateMachine.getWarningCollector(), statsCalculator)
+                .analyze(statement);
+        ExpressionAnalyzer analyzer = ExpressionAnalyzer.create(analysis, session, metadata, sqlParser, groupProvider, accessControl,
+                TypeProvider.empty(), warningCollector, CorrelationSupport.ALLOWED);
+
+        Type type = analyzer.analyze(expression, Scope.builder()
+                .withRelationType(RelationId.anonymous(), new RelationType())
+                .build());
+        Map<NodeRef<Expression>, Type> expressionTypes = ImmutableMap.<NodeRef<Expression>, Type>builder()
+                .put(NodeRef.of(expression), type)
+                .build();
+        ExpressionInterpreter interpreter = new ExpressionInterpreter(expression, metadata, session, expressionTypes);
+        return interpreter.evaluate();
```

Comment:
This could be simplified to:
```
Session session = stateMachine.getSession();
Map<NodeRef<Parameter>, Expression> parameterLookup = ParameterUtils.parameterExtractor(statement, parameters);
return ExpressionInterpreter.evaluateConstantExpression(
            expression,
            expectedType,
            metadata,
            session,
            accessControl,
            parameterLookup)
```
However, you need to provide the `expectedType`. To accomplish that, you should probably make a distinction between the two supported argument types (that is, string and interval) in the AST node.

Reply:
Thanks @kasiafi  - this is definitely a good suggestion.